runtime/ts/storage/synthetic-storage.ts

```diff
@@ -0,0 +1,132 @@
+// @
+// Copyright (c) 2018 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {assert} from '../../../platform/assert-web.js';
+import {StorageBase, StorageProviderBase} from './storage-provider-base.js';
+import {KeyBase} from './key-base.js';
+import {FirebaseStorage} from './firebase-storage.js';
+import {Id} from '../id.js';
+import {Type} from '../type.js';
+import {Manifest} from '../../manifest.js';
+
+enum Scope {
+  arc = 1  // target must be a storage key referring to a serialized manifest
+}
+
+enum Category {
+  handles = 1
+}
+
+// Format is 'synthetic://<scope>/<category>/<target>'
+class SyntheticKey extends KeyBase {
+  scope: Scope;
+  category: Category;
+  target: string;
+  constructor(key: string) {
+    super();
+    const match = key.match(/^synthetic:\/\/([^/]*)\/([^/]*)\/(.*)$/);
+    assert(match && match.length === 4, `invalid synthetic key: ${key}`);
+    this.scope = Scope[match[1]];
+    this.category = Category[match[2]];
+    this.target = match[3];
+    assert(this.scope, `invalid scope '${match[1]}' for synthetic key: ${key}`);
+    assert(this.category, `invalid category '${match[2]}' for synthetic key: ${key}`);
+  }
+
+  get protocol() {
+    return 'synthetic';
+  }
+
+  childKeyForHandle(id): SyntheticKey {
+    assert(false, 'childKeyForHandle not supported for synthetic keys');
+    return null;
+  }
+
+  toString() {
+    return `${this.protocol}://${Scope[this.scope]}/${Category[this.category]}/${this.target}`;
+  }
+}
+
+// TODO: unhack this
+function isFirebaseKey(key) {
+  return key && key.startsWith('firebase:');
+}
+
+export class SyntheticStorage extends StorageBase {
+  private firebaseStorage: FirebaseStorage;
+
+  constructor(arcId: Id, firebaseStorage) {
+    super(arcId);
+    this.firebaseStorage = firebaseStorage;
+  }
+
+  async construct(id: string, type: Type, keyFragment: string) : Promise<SyntheticCollection> {
+    throw new Error('cannot construct synthetic storage providers; use connect');
+  }
+
+  async connect(id: string, type: Type, key: string) : Promise<SyntheticCollection> {
+    // TODO: fix the API so we don't need this workaround
```

Comment:
instead: // TODO: add handle type to type system.

Reply:
Done.